rumqttc/src/v5/mqttbytes/mod.rs

```diff
@@ -32,61 +36,26 @@ pub fn qos(num: u8) -> Option<QoS> {
     }
 }
 
-/// Checks if a topic or topic filter has wildcards
-pub fn has_wildcards(s: &str) -> bool {
-    s.contains('+') || s.contains('#')
-}
-
-/// Checks if a topic is valid
-pub fn valid_topic(topic: &str) -> bool {
-    // topic can't contain wildcards
-    if topic.contains('+') || topic.contains('#') {
-        return false;
-    }
-
-    true
-}
-
-/// Checks if the filter is valid
-///
-/// <https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718106>
-pub fn valid_filter(filter: &str) -> bool {
-    if filter.is_empty() {
-        return false;
-    }
-
-    // rev() is used so we can easily get the last entry
-    let mut hirerarchy = filter.split('/').rev();
-
-    // split will never return an empty iterator
-    // even if the pattern isn't matched, the original string will be there
-    // so it is safe to just unwrap here!
-    let last = hirerarchy.next().unwrap();
-
-    // only single '#" or '+' is allowed in last entry
-    // invalid: sport/tennis#
-    // invalid: sport/++
-    if last.len() != 1 && (last.contains('#') || last.contains('+')) {
-        return false;
-    }
-
-    // remaining entries
-    for entry in hirerarchy {
-        // # is not allowed in filter except as a last entry
-        // invalid: sport/tennis#/player
-        // invalid: sport/tennis/#/ranking
-        if entry.contains('#') {
-            return false;
-        }
-
-        // + must occupy an entire level of the filter
-        // invalid: sport+
-        if entry.len() > 1 && entry.contains('+') {
-            return false;
-        }
-    }
-
-    true
+pub(crate) fn validate_topic_name_and_alias(
+    topic: impl AsRef<str>,
+    properties: &Option<PublishProperties>,
+) -> bool {
+    let topic = topic.as_ref();
+    let is_topic_empty = topic.is_empty();
+    // The topic alias is considered valid only if it is greater than zero.
+    // If it is not supplied, it is still considered valid because in that
+    // case, the validity depends on the topic name itself.
+    let is_topic_valid = valid_topic(topic);
+    let is_alias_given = properties
+        .as_ref()
+        .map_or(false, |p| p.topic_alias.is_some());
+    let is_alias_valid = properties
+        .as_ref()
+        .map_or(false, |p| p.topic_alias.map_or(false, |a| a > 0));
+
+    (is_topic_valid || is_topic_empty)
+        && (!is_topic_empty || is_alias_given)
+        && (!is_alias_given || is_alias_valid)
```

Comment:
imo this is harder to read in comparison to if...else branch, wdyt @swanandx?

Reply:
I did some investigation. It seems that apart from the length of the function, the assembly for if-else version of `validate_topic_name_and_alias` remains the same. How should we proceed?

https://godbolt.org/z/6q4qd75nY